build-tools/src/main/java/org/elasticsearch/gradle/plugin/GenerateTestBuildInfoTask.java

```diff
@@ -0,0 +1,294 @@
+ */
+
+package org.elasticsearch.gradle.plugin;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.TaskAction;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ModuleVisitor;
+import org.objectweb.asm.Opcodes;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.ZipEntry;
+
+/**
+ * This task generates a file with a class to module mapping
+ * used to imitate modular behavior during unit tests so
+ * entitlements can lookup correct policies.
+ */
+public abstract class GenerateTestBuildInfoTask extends DefaultTask {
+
+    public static final String DESCRIPTION = "generates plugin test dependencies file";
+
+    public GenerateTestBuildInfoTask() {
+        setDescription(DESCRIPTION);
+    }
+
+    @Input
+    @Optional
+    public abstract Property<String> getModuleName();
+
+    @Input
+    public abstract Property<String> getComponentName();
+
+    @InputFiles
+    public abstract Property<FileCollection> getCodeLocations();
+
+    @Input
+    public abstract Property<String> getOutputFileName();
+
+    @OutputDirectory
+    public abstract DirectoryProperty getOutputDirectory();
+
+    @TaskAction
+    public void generatePropertiesFile() throws IOException {
+        Map<String, String> classesToModules = buildClassesToModules();
+
+        Path outputDirectory = getOutputDirectory().get().getAsFile().toPath();
+        Files.createDirectories(outputDirectory);
+        Path outputFile = outputDirectory.resolve(getOutputFileName().get());
+
+        try (var writer = Files.newBufferedWriter(outputFile, StandardCharsets.UTF_8)) {
+            writer.write("{\n");
+
+            writer.write("    \"name\": \"");
+            writer.write(getComponentName().get());
+            writer.write("\",\n");
+
+            writer.write("    \"locations\": [\n");
+            if (classesToModules.isEmpty() == false) {
+                StringBuilder sb = new StringBuilder();
+                for (Map.Entry<String, String> entry : classesToModules.entrySet()) {
+                    sb.append("        {\n");
+                    sb.append("            \"class\": \"");
```

Comment:
nit: I find "class" to be overly generic here, there isn't one class for the entire module/jar. IMO this should be "representative_class" or something similar, though I don't feel strongly, just a thought.

Reply:
@prdoyle changes this to `representativeClass`